From 32c0009abcd7c86492d6f0460331f0ff6efa8e7a Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Tue, 4 Jun 2019 20:35:39 +0200
Subject: [PATCH 1/1] efi_loader: call event function at event TPL

Event notification functions have to be executed at the event's task
priority level.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 lib/efi_loader/efi_boottime.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/lib/efi_loader/efi_boottime.c b/lib/efi_loader/efi_boottime.c
index 7d1d6e9213..a67c58fbb6 100644
--- a/lib/efi_loader/efi_boottime.c
+++ b/lib/efi_loader/efi_boottime.c
@@ -177,13 +177,19 @@ const char *__efi_nesting_dec(void)
 static void efi_queue_event(struct efi_event *event, bool check_tpl)
 {
 	if (event->notify_function) {
+		efi_uintn_t old_tpl;
+
 		event->is_queued = true;
 		/* Check TPL */
 		if (check_tpl && efi_tpl >= event->notify_tpl)
 			return;
 		event->is_queued = false;
+		/* Events must be executed at the event's TPL */
+		old_tpl = efi_tpl;
+		efi_tpl = event->notify_tpl;
 		EFI_CALL_VOID(event->notify_function(event,
 						     event->notify_context));
+		efi_tpl = old_tpl;
 	} else {
 		event->is_queued = false;
 	}
-- 
2.20.1

